python/pyarrow/tests/test_udf.py

```diff
@@ -0,0 +1,447 @@
+
+    # doc with no decription
+    func_doc = {
+        "summary": "test summary"
+    }
+
+    expected_expr = "Function doc must contain a description"
+
+    with pytest.raises(ValueError, match=expected_expr):
+        register_scalar_function("no_desc", func_doc, in_types,
+                                 out_type, unary_scalar_function)
+
+    # doc with empty dictionary
+    func_doc = {}
+    expected_expr = r"Function doc must contain a summary,"
+    with pytest.raises(ValueError, match=expected_expr):
+        register_scalar_function("empty_dictionary", func_doc, in_types,
+                                 out_type, unary_scalar_function)
+
+
+def test_non_uniform_input_udfs():
+
+    def unary_scalar_function(scalar1, array1, scalar2):
+        coeff = pc.call_function("add", [scalar1, scalar2])
+        return pc.call_function("multiply", [coeff, array1])
+
+    in_types = {"scalar1": InputType.scalar(pa.int64()),
+                "scalar2": InputType.array(pa.int64()),
+                "scalar3": InputType.scalar(pa.int64()),
+                }
+    func_doc = {
+        "summary": "multi type udf",
+        "description": "desc"
+    }
+    register_scalar_function("multi_type_udf", func_doc,
+                             in_types,
+                             pa.int64(), unary_scalar_function)
+
+    res = pc.call_function("multi_type_udf",
+                           [pa.scalar(10), pa.array([1, 2, 3]), pa.scalar(20)])
+
+    assert res == pa.array([30, 60, 90])
+
+
+def test_nullary_functions():
+
+    def gen_random():
+        import random
+        val = random.randint(0, 10)
+        return pa.scalar(val)
+
+    func_doc = {
+        "summary": "random function",
+        "description": "generates a random value"
+    }
+
+    register_scalar_function("random_func", func_doc,
+                             {},
+                             pa.int64(), gen_random)
+
+    res = pc.call_function("random_func", [])
+    assert res.as_py() >= 0 and res.as_py() <= 10
+
+
+def test_output_datatype():
+    def add_one(array):
+        ar = pc.call_function("add", [array, 1])
+        ar = ar.cast(pa.int32())
+        return ar
+
+    func_name = "py_add_to_scalar"
+    in_types = {"array": InputType.array(pa.int64())}
+    out_type = pa.int64()
+    doc = {
+        "summary": "add function scalar",
+        "description": "add function"
+    }
+    register_scalar_function(func_name, doc,
+                             in_types, out_type, add_one)
+
+    func = pc.get_function(func_name)
+
+    assert func.name == func_name
+
+    expected_expr = "Expected output type, int64," \
+        + " but function returned type int32"
+
+    with pytest.raises(ValueError, match=expected_expr):
+        pc.call_function(func_name, [pa.array([20, 30])])
+
+
+def test_output_type():
+    def add_one(array):
+        ar = pc.call_function("add", [array, 1])
+        ar = ar.cast(pa.int32())
+        return ar[0].as_py()
```

Comment:
You can make this much simpler, e.g.:
```suggestion
        return 42
```